### ISSUE TITLE
"Compare Scopesets" does not expand scopes

### ISSUE THREAD

**Issue by djmitche** [taskcluster/taskcluster#1397]
**Describe the bug**
The scope comparison tool should expand any `assume:..` roles in the scopes and then compare the resulting, expanded scopesets.  It appears to just compare the given scopesets.

**To Reproduce**
1. Go to https://taskcluster-ui.herokuapp.com/auth/scopes/compare
1. Enter `assume:repo:github.com/taskcluster/monopacker:branch:*` in one box and `assume:repo:github.com/taskcluster/*` in the other.
1. Click on the button thing in the lower right
1. See just the two entered scopes in the resulting diff -- not the scopes that they expand to.

**Expected behavior**
Expansion of the scopes

**Comment by helfi92** [comment]
From my understanding of https://bugzilla.mozilla.org/show_bug.cgi?id=1444268#c11, the compare scopesets tool is not supposed to expand scopes. There's a PR open to fix the regression that was introduced in the view that expands scopes. Maybe that will help?

**Comment by djmitche** [comment]
Ah, that bug is at best ambiguous about it!  Could we consider this a feature request, then?  Perhaps an "expand" button on each scope-entry widget that will expand it?

**Comment by helfi92** [comment]
> Could we consider this a feature request, then?

Sure thing!

**Comment by ANURADHAJHA99** [comment]
@djmitche Can you please assign the above issue to me? Thank you.

**Comment by djmitche** [comment]
Sure thing!  This one will be a good bit more challenging.  Look at the "scope expander" tool, and check the docs about roles and scopes for some background.

**Comment by ANURADHAJHA99** [comment]
@djmitche Sure, That's what I am currently doing, trying to understand the tool. Thank you very much.

**Comment by ANURADHAJHA99** [comment]
@djmitche I used the expander tool. However even after reading the docs, I am still unable to comprehend how did it expand to so many scopes. How are the relation of one scope defined with another scope? I am very sorry but can you please clear my doubt?

**Comment by djmitche** [comment]
[This page](https://docs.taskcluster.net/docs/reference/platform/auth/scopes) and the few pages that follow it should help explain.  Basically, a scope that starts with `assume:` can "expand" into a lot more scopes, and the scope expander tool is a way for users to see this in action.  You can see the roles that are defined in https://taskcluster-ui.herokuapp.com/auth/roles.

**Comment by ANURADHAJHA99** [comment]
@djmitche So our main task is to introduce the aspect of "scope expander tool" in "Scope comparison tool"?

**Comment by djmitche** [comment]
Yes, as an option (so a button for each text area that will expand the scopes in that text area)

**Comment by ANURADHAJHA99** [comment]
Okay, Thank you so much for the help. I got it now. I am working on it!!

**Comment by ANURADHAJHA99** [comment]
@djmitche I was going through the taskcluster repository and tried to understand the codebase of 'RoleScopesTable' and other files in the components directory. I am however stuck, about the expandScope function. Is it available in the 'utils' directory? I wanted to see the function to get an idea about how exactly the expansion of scopes takes place. Also while comparing the two scopes there was a placeholder in the two input box "new-scope:for-something:*". While inserting the typing cursor in the box, however, the placeholder seemed to not vanish which made me assume it is permanent or a scope from a new user perspective. Can I fix the placeholder, such that it gets removes when the text cursor is placed, which is a more common behavior of a generic placeholder? Thank you very much for clearing my doubts.

**Comment by djmitche** [comment]
I'm glad you're working to figure this out!  The expandScopes function is an [API method on the auth service](https://docs.taskcluster.net/docs/reference/platform/auth/api#expandScopes).  You can find the code [here](https://github.com/taskcluster/taskcluster/blob/master/services/auth/src/scoperesolver.js) but it's probably not an especially enlightening read -- it's heavily optimized so that it can expand scopes quickly, which makes the meaning of the code a lot less clear.  The first time I read that code I did so with several pieces of notebook paper and drew out a bunch of tries to try to understand it.  Probably the best explanation is in [the docs](https://docs.taskcluster.net/docs/reference/platform/auth/roles).

Fixing the placeholder seems a good idea!

**Comment by ANURADHAJHA99** [comment]
@djmitche I was working on the scopesetComparison index.jsx and I finally understood how the scopesetExpander.jsx works. However, I had a doubt, can we use the "ScopeExpander class" in scopesetComparison.jsx by importing it and add an expander button on the bottom left? That's how I am thinking about forming the solution. Please correct me if I am wrong. Thank you for helping me.

**Comment by ANURADHAJHA99** [comment]
@djmitche I have added the "ScopeExpander button" as required, however, i am having difficulty styling the button. Actually I have to position the button in the bottom left which is creating some problems, 
@withStyles(theme => ({
  actionButton: {
  ...theme.mixins.fab,
  },
this is the piece of code for styling the button, I have tried googling but it is not helping out, can you please guide me? Thank you very much for helping.

**Comment by ANURADHAJHA99** [comment]
@helfi92 Can you please guide me with the above problems? Thank you for answering my queries.

**Comment by djmitche** [comment]
> can we use the "ScopeExpander class"

I don't see a class with that name, so I'm not sure how to answer.

**Comment by helfi92** [comment]
@ANURADHAJHA99 We use [CSS-in-JS](https://material-ui.com/styles/basics/) to for styling. See the examples there to see how they style buttons.

> Actually I have to position the button in the bottom left which is creating some problems,
@withStyles(theme => ({
actionButton: {
...theme.mixins.fab,
},
this is the piece of code for styling the button, I have tried googling but it is not helping out, can you please guide me? Thank you very much for helping.

The definition of theme.mixins.fab is defined in theme.js https://github.com/taskcluster/taskcluster/blob/7dbcbba2de43bf01f8dc5775b97ea914906787cc/ui/src/theme.js#L186-L193

Does that help?
